apps/srt-live-transmit.cpp

```diff
@@ -572,7 +590,7 @@ int main( int argc, char** argv )
                     std::shared_ptr<bytevector> pdata = dataqueue.front();
                     if (!tar.get() || !tar->IsOpen()) {
                         lostBytes += (*pdata).size();
-                    } else if (!tar->Write(*pdata)) {
+                    } else if (!tar->Write(pdata->data(), pdata->size(), out_stats)) {
```

Comment:
What's the reason of having data/size passed separately? It doesn't make a difference, except that you have to pass two of them.

Also, the `Read` and `Write` functions are predicted to implement a single entry point for reading and writing; adding more virtual functions demolishes the design. I personally think that there's nothing wrong in having this `out_stats` parameter written into a global variable, like many other parameters.

Reply:
This was done mainly because of the file transmit uses the same preallocated buffer in a loop.
```
vector<char> buf(cfg.chunk_size);
```
While `live-transmit` allocates a new buffer every time, which is not very good.
```
std::shared_ptr<bytevector> pdata(
                            new bytevector(cfg.chunk_size));
```
That is why passing a container is not a good approach, because the size of the container may be larger, than the size of the valid data contained.

> Also, the Read and Write functions are predicted to implement a single entry point for reading and writing; adding more virtual functions demolishes the design.

I would then delete the function with the container passed.

> I personally think that there's nothing wrong in having this out_stats parameter written into a global variable

Global variables are hard to follow, when passed from one module to another via several nested functions.

Reply:
If you leave just one version of Read and Write functions, this should be good enough.